android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/MainActivity.java

```diff
@@ -83,27 +80,93 @@ public boolean onOptionsItemSelected(MenuItem item) {
     return super.onOptionsItemSelected(item);
   }
 
+  /**
+   * Ensures Bluetooth is available on the beacon and it is enabled. If not,
+   * displays a dialog requesting user permission to enable Bluetooth.
+   */
+  private void ensureBluetoothAndLocationIsEnabled(BluetoothAdapter bluetoothAdapter) {
+    mCheckingPermissions = true;
+    if (!bluetoothAdapter.isEnabled()) {
+      Intent enableBtIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
+      startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT);
+      return;
+    }
+    ensureLocationPermissionIsEnabled();
+  }
+
+  @Override
+  protected void onActivityResult (int requestCode, int resultCode, Intent data) {
+    Log.d(TAG, "onActivityResult");
+    if (requestCode == REQUEST_ENABLE_BT && resultCode == -1) {
+      ensureLocationPermissionIsEnabled();
+      return;
+    }
+    Toast.makeText(this, getString(R.string.bt_on), Toast.LENGTH_LONG).show();
+    finish();
+  }
+
   @Override
   protected void onResume() {
     super.onResume();
-    Log.d(TAG, "resumed MainActivity");
-    BluetoothManager btManager = (BluetoothManager) getSystemService(BLUETOOTH_SERVICE);
-    BluetoothAdapter btAdapter = btManager != null ? btManager.getAdapter() : null;
-    if (btAdapter == null) {
-      Toast.makeText(getApplicationContext(),
-              R.string.error_bluetooth_support, Toast.LENGTH_LONG).show();
-      finish();
+    if(!checkingPermissions){
+      Log.d(TAG, "resumed MainActivity");
+      BluetoothManager btManager = (BluetoothManager) getSystemService(BLUETOOTH_SERVICE);
+      BluetoothAdapter btAdapter = btManager != null ? btManager.getAdapter() : null;
+      if (btAdapter == null) {
+        Toast.makeText(getApplicationContext(),
+            R.string.error_bluetooth_support, Toast.LENGTH_LONG).show();
+        finish();
+        return;
+      }
+      if (checkIfUserHasOptedIn()) {
+        Log.d(TAG, "checkingPermissions");
+        ensureBluetoothAndLocationIsEnabled(btAdapter);
+        
```

Comment:
It looks like this line has some trailing whitespace, could you remove it here and check the rest of the file?

Making whitespace visible in Sublime Text makes this easier, add this to your settings:
  "draw_white_space": "all"